swift/ql/lib/codeql/swift/security/CleartextStorageDatabaseExtensions.qll

```diff
@@ -122,14 +122,21 @@ private class CleartextStorageDatabaseEncryptionSanitizer extends CleartextStora
 
 /**
  * An additional taint step for cleartext database storage vulnerabilities.
- * Needed until we have proper content flow through arrays.
  */
 private class CleartextStorageDatabaseArrayAdditionalTaintStep extends CleartextStorageDatabaseAdditionalTaintStep {
   override predicate step(DataFlow::Node nodeFrom, DataFlow::Node nodeTo) {
+    // needed until we have proper content flow through arrays.
     exists(ArrayExpr arr |
       nodeFrom.asExpr() = arr.getAnElement() and
       nodeTo.asExpr() = arr
     )
+    or
+    // if an object is sensitive, its fields are always sensitive
+    // (this is needed because the sensitive data sources are in a sense
+    //  approximate; for example we might identify `passwordBox` as a source,
+    //  whereas it is more accurate to say that `passwordBox.textField` is the
+    //  true source).
+    nodeTo.asExpr().(LookupExpr).getBase() = nodeFrom.asExpr()
```

Comment:
This basically introduces total field/object conflation for this query which may lead to a lot of FPs. Could we add it only at the sinks instead? Would that still catch the TPs we want to catch?

Reply:
I'm also a bit nervous about this, but I don't think adding it at the sinks is enough (maybe at the sources???).  I'm fairly sure I _can_ replace `LookupExpr` with something a bit narrower, but that will still be 'total conflation' in a smaller class of things.  I'll experiment a bit more...

Reply:
(we will want this new taint step on the other sensitive data queries as well when we're done here, so it's worth a bit of time getting it right...)

Reply:
I've narrowed the additional taint step from all `LookupExpr`s to just `MemberRefExpr`, and added a couple of additional test cases while I was thinking about this stuff.

I don't think a solution that works at the sinks can be good enough here.  I'm more interested in sources right now.  Consider:
```
var a = myPasswordBox
var b = a.textField
db.storedText = b
```
In this case the heuristic has determined `myPasswordBox` is sensitive, but *really* the sensitive data is `myPasswordBox.textField`.  What we really want is for all content of `myPasswordBox` on the first line to be considered tainted, then track where it goes.  But as far as I'm aware we don't have a tool for that (do we?), so we allow flow from `a` to `a.textField` instead.

What kinds of false positives worry you most?  I can imagine cases where a field of something we determine to be sensitive is actually not sensitive, but I think they will be unusual.  It's actually quite hard to think of examples, but I came up with:
- `password.length` should probably be sensitive IMO.
- `bankaccount.providerName` seems only a little bit sensitive, compared to `bankaccount.number` say.
- `mobilePhone.colour` wouldn't really be sensitive at all, though I think the real issue here is that matching `mobile` on its own is already dubious (I have an issue open to fix this).  Once that's done, the example would be `mobileNumber.colour` and that is nonsensical.  The idea here is that the sensitive data heuristic should be labelling fairly specific things, for which all fields would likely be sensitive.

Reply:
I can see that having the additional step only at the sink is going to be limiting, indeed.

> But as far as I'm aware we don't have a tool for that (do we?),

Correct, we don't have such a tool. There's no way to insert such "implicit stores" similar to how we can insert "implicit reads".

> What kinds of false positives worry you most? I can imagine cases where a field of something we determine to be sensitive is actually not sensitive, but I think they will be unusual.

I haven't thought too carefully about it yet. I was thinking of strange cases like `mobilePhone.colour`, but when you say
> The idea here is that the sensitive data heuristic should be labelling fairly specific things, for which all fields would likely be sensitive.

I'm much less concerned about this additional step 👍.